_posts/2022-04-24-ui-architecture.md

```diff
@@ -0,0 +1,264 @@
+</details>
+
+So what does Xilem do instead? The view tree is also parameterized on the *app state,* which can be any type. This idea is an evolution of Druid's existing architecture, which also offers mutable access to app state to UI callbacks, but removes some of the limitations. In particular, Druid requires app state to be clonable and diffable, a stumbling block for many new users.
+
+When an event is generated, it is annotated with the path of the UI element that originated it. In the case of the button, `[1, 3]`. We retain the view tree for a fairly short time, but long enough to do event propagation. The event is dispatched from the *root,* and each successive node strips off one id from the id path, choosing a child, and traversing to that child. In this case, that's simple, the enclosing stack just chooses the button. This event propagation also passes down mutable access to the app state.
+
+Given that setup, running the closure is fairly straightforward. The event is dispatched to the stack, which uses the id to recognize that it needs to be routed to its second child, and that button is the target. It runs the callback on the mutable app state, which in this case just increments `count`.
+
+### Re-rendering
+
+After clicking the button and running the callback, the app state consists of the number 1, formerly 0. The app logic function is run, producing a new view tree, and this time the string value is "Count: 1" rather than "Count: 0". The challenge is then to update the widget tree with the new data.
+
+As is completely standard in declarative UI, it is done by diffing the old view tree against the new one, in this case calling the `rebuild` method on the `View` trait. This method compares the data, updates the associated widget if there are any changes, and also traverses into children.
+
+In the simplest case, the app builds the full view tree, and that is diffed in full against the previous version. However, as UI scales, this would be inefficient, so there are *other* mechanisms to do finer grained change propagation, as described below.
+
+## Components
+
+The above is the basic architecture, enough to get started. Now we will go into some more advanced techniques.
+
+It would be very limiting to have a single "app state" type throughout the application, and require all callbacks to express their state mutations in terms of that global type. So we won't do that.
+
+The main tool for stitching together components is the `Adapt` view node. This node is so named because it adapts between one app state type and another, using a closure that takes mutable access to the parent state, and calls into a child (through a "thunk") with a mutable reference to the child state.
```

Comment:
You haven't defined what a 'component' is. Previous to this the reader knows only about views, widgets, and view state. Is a component just another term for a view? This section is titled 'components' but talks mostly about the adapt node (view?).